### ISSUE TITLE
Updating Azure LoadBalancer NSG source ranges does not work

### ISSUE THREAD

**Issue by rocketraman** [Azure/acs-engine#423]
When exposing a service with type LoadBalancer, the Azure load balancer and network security group is automatically updated.

However, the generated Inbound Rules in the network security group assume that the Source is "Internet". Is there a way to specify, perhaps via Service meta-data, that the service should be exposed with a different Source e.g. source tag like `VirtualNetwork`, a CIDR block, or Any?

On a related note, the fact that the resource group is created with a Vnet using `10.0.0.0/8` limits the ability to communicate via gateways to other VirtualNetworks that were previously created in the same network space. I see several issues about supporting the ability to change the Kubernetes subnet but it seems like a very finicky and manual process as of now.

**Comment by colemickens** [comment]
RE: LB NSG Source: Is this what you're after: https://github.com/kubernetes/kubernetes/pull/36696 ?

As for changing the CIDR spaces, see this PR: https://github.com/Azure/acs-engine/pull/191 It may unblock you before we can fully address the issue.

**Comment by rocketraman** [comment]
> RE: LB NSG Source: Is this what you're after: kubernetes/kubernetes#36696 ?

Not sure... I tried both a `service.beta.kubernetes.io/load-balancer-source-ranges` annotation as well as a `loadBalancerSourceRanges` entry in the spec of my nginx-ingress-controller service. Neither seemed to have any affect on the load balancer or NSG definition.

> As for changing the CIDR spaces, see this PR: #191 It may unblock you before we can fully address the issue.

I see the PR, but I also see that it is currently out-of-date. Not sure I want to mess with the setup to this extent. Is there a workaround I can apply in the meantime? Not being able to have two-way communication between a Kubernetes resource group and any other resource groups anywhere in the 10.0.0.0/8 address space seems like a pretty big issue to me -- how are other people working around it? Can Azure VNET gateways use NAT addressing?

**Comment by colemickens** [comment]
*RE: Source Ranges*: That commit is in `1.5.0`. You're not running on an old `1.4.x` cluster by chance, are you?

I don't know if VNet gateways can NAT, but I've not heard of it previously. Other folks are emulating what is in that PR by editting the template output from ACS-Engine and changing the ranges, or shrinking them, etc. I've not heard of a better workaround so far, nor do I have any other suggestions to make unfortunately.

**Comment by rocketraman** [comment]
> RE: Source Ranges: That commit is in 1.5.0. You're not running on an old 1.4.x cluster by chance, are you?

I'm on 1.5.3. After modifying the resource json and applying it, I see the annotation/spec information in the Kubernetes dashboard. Does Kubernetes / Azure perhaps pick up the source ranges information only on first apply of the controller definition, and miss it for updates?  I'll try removing and recreating the controller.

> I don't know if VNet gateways can NAT, but I've not heard of it previously. Other folks are emulating what is in that PR by editting the template output from ACS-Engine and changing the ranges, or shrinking them, etc. I've not heard of a better workaround so far, nor do I have any other suggestions to make unfortunately.

All right, I'll give it a shot.

**Comment by rocketraman** [comment]
> Does Kubernetes / Azure perhaps pick up the source ranges information only on first apply of the controller definition, and miss it for updates?

Indeed, this seems to be a bug. If I delete the service and recreate it, the `loadBalancerSourceRanges` from the ingress service is picked up and applied to the NSG. However, If I simply modify an existing service and update `loadBalancerSourceRanges`, no change is made to the Azure NSG. This is true whether or not the existing service had `loadBalancerSourceRanges` defined or not.

**Comment by colemickens** [comment]
Please file that upstream on the Kubernetes repo and tag `brendandburns` in it. I'm not surprised there is a gap around updating an existing service.

Also, with kubernetes-dashboard, the addon manager may be trying to undo your changes. Are you sure they persist after editting the existing service, or do your changes get undone?

Finally, you might be better off creating a new service that reverse proxies to the dashboard. This is what I generally do, and then I can do authentication in that reverse proxy.

**Comment by rocketraman** [comment]
> Please file that upstream on the Kubernetes repo and tag brendandburns in it. I'm not surprised there is a gap around updating an existing service.

There seems to be one similar issue already (https://github.com/kubernetes/kubernetes/issues/34218) though its not exactly the same -- in that issue the reporter was able to get the AWS SecurityGroup to update by changing not just the `loadBalancerSourceRanges` but also something else, like an annotation. In my case, changing an annotation still does not cause `loadBalancerSourceRanges` to be picked up.

I created new issue https://github.com/kubernetes/kubernetes/issues/43608 as requested.

> Also, with kubernetes-dashboard, the addon manager may be trying to undo your changes. Are you sure they persist after editting the existing service, or do your changes get undone?

No, I am editing via `kubectl apply` and the changes seem to stick. They are just never applied to the NSG.

> Finally, you might be better off creating a new service that reverse proxies to the dashboard. This is what I generally do, and then I can do authentication in that reverse proxy.

This is a good idea but I'm a bit confused about the context? Is this related to the current issue or is it just useful "drive-by" information? :-)

**Comment by colemickens** [comment]
Thank you for creating the issue and clarifying that your changes to dashboard yaml stick. We'll take a look as soon as possible.

>Is this related to the current issue or is it just useful "drive-by" information? :-)

Hopefully useful drive-by information :).

**Comment by anhowe** [comment]
Closing since this is covered by upstream issue.  Please re-open, if you think changes are additionally required in ACS-Engine to make this work.
